### PR TITLE
Support helm chart replication from helmhub

### PR DESCRIPTION
**Pull request by mmpei**
Signed-off-by: peimingming <peimingming@corp.netease.com>

it will invoke getChartDetail 3-5 times to make it compatible with the chart interface.
so this will degrade some performance.
@ywk253100 @steven-zou @cd1989 @kofj

### CODE REVIEW

**Comment by kofj** [comment]
@mmpei Please fix: https://travis-ci.org/goharbor/harbor/jobs/556101182

File: src/replication/adapter/helmhub/chart_registry.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	for _, v := range versionList.Data {
+		if v.Attributes.Version == version {
+			return a.download(v)
+		}
+	}
+	return nil, nil
```
Comment (ywk253100) [inline src/replication/adapter/helmhub/chart_registry.go]: Should return error here to indicates that the chart is not found

File: src/replication/adapter/helmhub/client.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+// Client is a client to interact with HelmHub
+type Client struct {
+	client *http.Client
+}
+
+// NewClient creates a new HelmHub client.
+func NewClient(registry *model.Registry) *Client {
+	return &Client{
+		client: &http.Client{
+			Transport: util.GetHTTPTransport(false),
```
Comment (ywk253100) [inline src/replication/adapter/helmhub/client.go]: Should read the `insecure` property of `registry` to set the parameter to `GetHTTPTransport`

File: src/replication/adapter/helmhub/adapter.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	}, nil
+}
+
+func (a *adapter) PrepareForPush(resources []*model.Resource) error {
+	return errors.New("not supported")
+}
+
+// HealthCheck checks health status of a registry
+func (a *adapter) HealthCheck() (model.HealthStatus, error) {
+	return model.Healthy, nil
```
Comment (ywk253100) [inline src/replication/adapter/helmhub/adapter.go]: Check the connection

File: src/replication/adapter/helmhub/chart_registry.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		if err = filter.DoFilter(&repositories); err != nil {
+			return nil, err
+		}
+	}
+
+	for _, repository := range repositories {
+		versionList, err := a.client.fetchChartDetail(repository.Name)
+		if err != nil {
+			log.Errorf("fetch chart detail: %v", err)
+			continue
```
Comment (ywk253100) [inline src/replication/adapter/helmhub/chart_registry.go]: Should return the error directly, two reasons:
1. If we keep the process logic even if got an error already, the users don't know it in the portal, the replication is marked as succeed, but actually it is not.
2.  To keep consistence with other adapters

File: src/replication/adapter/helmhub/client.go (hunk 2 above)
Comment (mmpei) [reply src/replication/adapter/helmhub/client.go]: the helm hub URL is fixed. and it is from offical. so secure should be checked. also refering to dockerhub.

File: src/replication/adapter/helmhub/adapter.go (hunk 3 above)
Comment (mmpei) [reply src/replication/adapter/helmhub/adapter.go]: ok

File: src/replication/adapter/helmhub/chart_registry.go (hunk 4 above)
Comment (mmpei) [reply src/replication/adapter/helmhub/chart_registry.go]: ok

File: src/replication/adapter/helmhub/chart_registry.go (hunk 1 above)
Comment (mmpei) [reply src/replication/adapter/helmhub/chart_registry.go]: ok

File: src/replication/adapter/helmhub/client.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	err = json.Unmarshal(body, list)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal chart detail response error: %v", err)
+	}
+
+	return list, nil
+}
+
+func (c *Client) checkHealthy() error {
+	request, err := http.NewRequest(http.MethodGet, baseURL, nil)
```
Comment (steven-zou) [inline src/replication/adapter/helmhub/client.go]: Shall we check the` index.yaml`  service endpoint? Seems the base URL is the root URL of the website?

File: src/replication/adapter/helmhub/client.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNotFound {
```
Comment (steven-zou) [inline src/replication/adapter/helmhub/client.go]: How about changing to the following code?
```go
if resp.StatusCode == http.StatusNotFound {
    return nil, ErrHTTPNotFound
}

if resp.StatusCode != http.StatusOK {
   return nil, fmt.Errorf("fetch chart detail error %d: %s", resp.StatusCode, string(body))
}
```

File: src/replication/adapter/helmhub/chart_registry.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+	for _, v := range versionList.Data {
+		if v.Attributes.Version == version {
+			return a.download(v)
+		}
+	}
+	return nil, errors.New("chart not found")
+}
+
+func (a *adapter) download(version *chartVersion) (io.ReadCloser, error) {
+	if version.Attributes.URLs == nil || len(version.Attributes.URLs) == 0 || len(version.Attributes.URLs[0]) == 0 {
```
Comment (steven-zou) [inline src/replication/adapter/helmhub/chart_registry.go]: `if version.Attributes.URLs == nil || len(version.Attributes.URLs) == 0 || len(version.Attributes.URLs[0]) == 0` ==>
`
if len(version.Attributes.URLs) == 0 || len(version.Attributes.URLs[0]) == 0
`

File: src/replication/adapter/helmhub/chart_registry.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+				},
+			})
+		}
+	}
+	return resources, nil
+}
+
+func (a *adapter) ChartExist(name, version string) (bool, error) {
+	versionList, err := a.client.fetchChartDetail(name)
+	if err != nil && err == ErrHTTPNotFound {
```
Comment (steven-zou) [inline src/replication/adapter/helmhub/chart_registry.go]:
```go

if err!=nil {
  if err == ErrHTTPNotFound {
    err = nil
  }

  return false, err
}

File: src/replication/adapter/helmhub/adapter.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func newAdapter(registry *model.Registry) (*adapter, error) {
+	return &adapter{
+		registry: registry,
+		client:   NewClient(registry),
+	}, nil
+}
+
+func (a *adapter) Info() (*model.RegistryInfo, error) {
```
Comment (steven-zou) [inline src/replication/adapter/helmhub/adapter.go]: From the implementation code, we can see that the helm hub adapter does not support push mode. Have we considered declaring such capability too?

@mmpei @ywk253100

File: src/replication/adapter/helmhub/chart_registry.go (hunk 8 above)
Comment (mmpei) [reply src/replication/adapter/helmhub/chart_registry.go]: looks strange.  but can change to the nested struct.

File: src/replication/adapter/helmhub/client.go (hunk 5 above)
Comment (mmpei) [reply src/replication/adapter/helmhub/client.go]: i don't know what you mean. but i couldn't find any related API. and i want to use `/charts` as health URL at first, but response body is too large.

File: src/replication/adapter/helmhub/adapter.go (hunk 9 above)
Comment (mmpei) [reply src/replication/adapter/helmhub/adapter.go]: it seams that there are some restrictions in UI. maybe hard code now.
agree with @ywk253100

**Pull request merged**